src/EventStore.Core/Services/RedactionService.cs

```diff
@@ -0,0 +1,261 @@
+
+			envelope.ReplyWith(new RedactionMessage.SwitchChunkCompleted(SwitchChunkResult.Success));
+		}
+
+		private static bool IsUnsafeFileName(string fileName) {
+			// protect against directory traversal attacks
+			return fileName.Contains('/') || fileName.Contains('\\') || fileName.Contains("..");
+		}
+
+		private bool IsValidSwitchChunkRequest(string targetChunkFile, string newChunkFile, out TFChunk newChunk, out SwitchChunkResult failReason) {
+			newChunk = null;
+
+			if (IsUnsafeFileName(targetChunkFile)) {
+				failReason = SwitchChunkResult.TargetChunkFileNameInvalid;
+				return false;
+			}
+
+			if (IsUnsafeFileName(newChunkFile)) {
+				failReason = SwitchChunkResult.NewChunkFileNameInvalid;
+				return false;
+			}
+
+			int targetChunkNumber;
+			try {
+				targetChunkNumber = _db.Config.FileNamingStrategy.GetIndexFor(targetChunkFile);
+			} catch {
+				failReason = SwitchChunkResult.TargetChunkFileNameInvalid;
+				return false;
+			}
+
+			if (Path.GetExtension(newChunkFile) != NewChunkFileExtension) {
+				failReason = SwitchChunkResult.NewChunkFileNameInvalid;
+				return false;
+			}
+
+			if (!File.Exists(Path.Combine(_db.Config.Path, targetChunkFile))) {
+				failReason = SwitchChunkResult.TargetChunkFileNotFound;
+				return false;
+			}
+
+			var newChunkPath = Path.Combine(_db.Config.Path, newChunkFile);
+			if (!File.Exists(newChunkPath)) {
+				failReason = SwitchChunkResult.NewChunkFileNotFound;
+				return false;
+			}
+
+			TFChunk targetChunk;
+			try {
+				targetChunk = _db.Manager.GetChunk(targetChunkNumber);
+			} catch(ArgumentOutOfRangeException) {
+				failReason = SwitchChunkResult.TargetChunkExcessive;
+				return false;
+			}
+
+			if (Path.GetFileName(targetChunk.FileName) != targetChunkFile) {
+				failReason = SwitchChunkResult.TargetChunkInactive;
+				return false;
+			}
+
+			if (targetChunk.ChunkFooter is not { IsCompleted: true }) {
+				failReason = SwitchChunkResult.TargetChunkNotCompleted;
+				return false;
+			}
+
+			ChunkHeader newChunkHeader;
+			ChunkFooter newChunkFooter;
+			try {
+				using var fs = new FileStream(newChunkPath, FileMode.Open, FileAccess.Read, FileShare.ReadWrite);
+				try {
+					newChunkHeader = ChunkHeader.FromStream(fs);
+					fs.Seek(-ChunkFooter.Size, SeekOrigin.End);
+					newChunkFooter = ChunkFooter.FromStream(fs);
+				} catch {
+					failReason = SwitchChunkResult.NewChunkHeaderOrFooterInvalid;
+					return false;
+				}
+			} catch {
+				failReason = SwitchChunkResult.NewChunkOpenFailed;
+				return false;
+			}
+
+			if (newChunkHeader.ChunkStartNumber != targetChunk.ChunkHeader.ChunkStartNumber ||
+			    newChunkHeader.ChunkEndNumber != targetChunk.ChunkHeader.ChunkEndNumber) {
+				failReason = SwitchChunkResult.ChunkRangeDoesNotMatch;
+				return false;
+			}
+
+			if (!newChunkFooter.IsCompleted) {
+				failReason = SwitchChunkResult.NewChunkNotCompleted;
+				return false;
+			}
+
+			try {
+				newChunk = TFChunk.FromCompletedFile(
```

Comment:
could add a comment saying that the chunkmanager will re-open the file with its own settings. when i first saw this i was worried that this TFChunk would be added to the manager with a maxreadercount of 1

Reply:
fixed :+1: